File: backend/pdf-demo-vps/apps/server/clients/opshub-client.ts
/** OpsHub drop-in client (per fleet convention): error logs + heartbeat.
 * Never throws into the host app; buffers to a local file when OpsHub is down. */
import { appendFileSync, existsSync, readFileSync, rmSync } from "node:fs";

interface LogEntry {
  service: string;
  level: "error" | "critical";
  event: string;
  message: string;
  traceback?: string;
  request_id?: string;
  meta?: Record<string, unknown>;
}

const URL_BASE = process.env.OPSHUB_URL ?? "";
const KEY = process.env.OPSHUB_KEY ?? "";
const SERVICE = process.env.OPSHUB_SERVICE ?? "pdf-demo";
const BUFFER_FILE = "/tmp/opshub-buffer.ndjson";
const enabled = URL_BASE !== "";

let lastHeartbeat = 0;

async function post(path: string, body: unknown): Promise<boolean> {
  if (!enabled) return true;
  try {
    const res = await fetch(`${URL_BASE}${path}`, {
      method: "POST",
      headers: { "content-type": "application/json", "x-opshub-key": KEY },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(3000)
    });
    return res.ok;
  } catch {
    return false;
  }
}

export async function opshubRegister(): Promise<void> {
  await post("/api/register", { service: SERVICE, container_name: SERVICE, url_health: "/health" });
  await flushBuffer();
}

export function opshubError(event: string, err: unknown, meta?: Record<string, unknown>): void {
  const entry: LogEntry = {
    service: SERVICE,
    level: "error",
    event,
    message: String((err as Error)?.message ?? err).slice(0, 2000),
    traceback: (err as Error)?.stack?.slice(0, 8000),
    meta
  };
  void post("/api/log", [entry]).then(ok => {
    if (!ok) {
      try { appendFileSync(BUFFER_FILE, JSON.stringify(entry) + "\n"); } catch { /* never throw */ }
    }
  });
}

/** Call from a request hook; throttled to 1/min so每 request stays cheap. */
export function opshubHeartbeat(): void {
  const now = Date.now();
  if (now - lastHeartbeat < 60_000) return;
  lastHeartbeat = now;
  void post("/api/heartbeat", { service: SERVICE });
}

async function flushBuffer(): Promise<void> {
  if (!existsSync(BUFFER_FILE)) return;
  try {
    const lines = readFileSync(BUFFER_FILE, "utf8").trim().split("\n").filter(Boolean).map(l => JSON.parse(l));
    if (lines.length && (await post("/api/log", lines))) rmSync(BUFFER_FILE);
  } catch { /* best effort */ }
}

File: backend/agents-demo/clients/opshub_client.py
"""OpsHub drop-in client (Python). Errors + heartbeat; never breaks the host app.
Usage:
    from opshub_client import opshub_register, opshub_error, heartbeat_middleware
    await opshub_register()                       # on startup
    app.middleware("http")(heartbeat_middleware)  # FastAPI: activity heartbeat
    opshub_error("job_failed", exc, {"job": id})  # on ERROR/CRITICAL only
"""
from __future__ import annotations

import json
import os
import time
import traceback as tb
from typing import Any

import httpx

URL = os.environ.get("OPSHUB_URL", "")
KEY = os.environ.get("OPSHUB_KEY", "")
SERVICE = os.environ.get("OPSHUB_SERVICE", "unknown")
BUFFER = "/tmp/opshub-buffer.ndjson"
_last_hb = 0.0


async def _post(path: str, body: Any) -> bool:
    if not URL:
        return True
    try:
        async with httpx.AsyncClient(timeout=3) as c:
            r = await c.post(f"{URL}{path}", json=body, headers={"x-opshub-key": KEY})
            return r.is_success
    except Exception:  # noqa: BLE001
        return False


async def opshub_register() -> None:
    await _post("/api/register", {"service": SERVICE, "container_name": SERVICE, "url_health": "/health"})
    await _flush()


def opshub_error(event: str, err: Any, meta: dict[str, Any] | None = None) -> None:
    entry = {
        "service": SERVICE, "level": "error", "event": event,
        "message": str(err)[:2000],
        "traceback": "".join(tb.format_exception(err))[:8000] if isinstance(err, BaseException) else None,
        "meta": meta,
    }
    import asyncio

    async def send() -> None:
        if not await _post("/api/log", [entry]):
            try:
                with open(BUFFER, "a") as f:
                    f.write(json.dumps(entry) + "\n")
            except OSError:
                pass

    try:
        asyncio.get_running_loop().create_task(send())
    except RuntimeError:
        asyncio.run(send())


async def heartbeat_middleware(request, call_next):  # FastAPI/Starlette middleware
    global _last_hb
    now = time.time()
    if now - _last_hb > 60:
        _last_hb = now
        import asyncio
        asyncio.get_running_loop().create_task(_post("/api/heartbeat", {"service": SERVICE}))
    return await call_next(request)


async def _flush() -> None:
    if not os.path.exists(BUFFER):
        return
    try:
        with open(BUFFER) as f:
            lines = [json.loads(x) for x in f.read().strip().splitlines() if x]
        if lines and await _post("/api/log", lines):
            os.remove(BUFFER)
    except Exception:  # noqa: BLE001
        pass
